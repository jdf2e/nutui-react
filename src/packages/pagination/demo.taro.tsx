import React, { useState } from 'react'
import Taro from '@tarojs/taro'
import { Left, Right } from '@nutui/icons-react-taro'
import { Pagination, Cell } from '@/packages/nutui.react.taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import Header from '@/sites/components/header'

interface T {
  ce5c5446: string
  c38a08ef: string
  b840c88f: string
  a74a1fd4: string
}

const PaginationDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      ce5c5446: '基础用法',
      c38a08ef: '简单模式',
      b840c88f: '显示省略号',
      a74a1fd4: '自定义按钮',
    },
    'zh-TW': {
      ce5c5446: '基礎用法',
      c38a08ef: '簡單模式',
      b840c88f: '顯示省略號',
      a74a1fd4: '自定義按鈕',
    },
    'en-US': {
      ce5c5446: 'Basic usage',
      c38a08ef: 'Simple mode',
      b840c88f: 'Show ellipsis',
      a74a1fd4: 'Custom button',
    },
  })
  const [currentPage1, setCurrent1] = useState(1)
  const [currentPage2, setCurrent2] = useState(1)
  const [currentPage3, setCurrent3] = useState(1)
  const [currentPage4, setCurrent4] = useState(3)
  const pageChange1 = (v: any) => {
    const c = v
    setCurrent1(c)
  }
  const pageChange2 = (v: any) => {
    const c = v
    setCurrent2(c)
  }
  const pageChange3 = (v: any) => {
    const c = v
    setCurrent3(c)
  }
  const pageChange4 = (v: any) => {
    const c = v
    setCurrent4(c)
  }
  const pageNodeRender = (item: any) => {
    return <div>{item.number === 3 ? 'hot' : item.text}</div>
  }
  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated.ce5c5446}</h2>
        <Cell>
          <Pagination
            modelValue={currentPage1}
            totalItems="20"
            itemsPerPage="5"
            onChange={pageChange1}
          />
        </Cell>
        <h2>{translated.c38a08ef}</h2>
        <Cell>
          <Pagination
            modelValue={currentPage2}
            pageCount={12}
            mode="simple"
            onChange={pageChange2}
          />
        </Cell>
        <h2>{translated.b840c88f}</h2>
        <Cell>
          <Pagination
            modelValue={currentPage3}
            totalItems="125"
            showPageSize="3"
            ellipse
            onChange={pageChange3}
          />
        </Cell>
        <h2>{translated.a74a1fd4}</h2>
        <Cell>
          <Pagination
            modelValue={currentPage4}
            totalItems="500"
            showPageSize="5"
            onChange={pageChange4}
            prev={<Left />}
            next={<Right />}
            pageNodeRender={pageNodeRender}
          />
        </Cell>
      </div>
    </>
  )
}

export default PaginationDemo
