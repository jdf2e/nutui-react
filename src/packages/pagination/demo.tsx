import React, { useState } from 'react'
import { Left, Right } from '@nutui/icons-react'
import Pagination from '@/packages/pagination'
import Cell from '@/packages/cell'
import { useTranslate } from '../../sites/assets/locale'

interface T {
  basic: string
  simple: string
  lite: string
  ellipse: string
  custom: string
  uncontrolled: string
}

const PaginationDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      basic: '基础用法',
      simple: '简单模式',
      lite: '极简模式',
      ellipse: '显示省略号',
      custom: '自定义按钮',
      uncontrolled: '非受控方式',
    },
    'zh-TW': {
      basic: '基礎用法',
      simple: '簡單模式',
      lite: '极简模式',
      ellipse: '顯示省略號',
      custom: '自定義按鈕',
      uncontrolled: '非受控方式',
    },
    'en-US': {
      basic: 'Basic usage',
      simple: 'Simple mode',
      lite: 'lite Mode',
      ellipse: 'Show ellipsis',
      custom: 'Custom button',
      uncontrolled: 'Uncontrolled mode',
    },
  })
  const [currentPage1, setCurrent1] = useState(1)
  const [currentPage2, setCurrent2] = useState(1)
  const [currentPage3, setCurrent3] = useState(1)
  const [currentPage4, setCurrent4] = useState(3)
  const pageChange1 = (v: number) => {
    const c = v
    setCurrent1(c)
  }
  const pageChange2 = (v: number) => {
    const c = v
    setCurrent2(c)
  }
  const pageChange3 = (v: number) => {
    const c = v
    setCurrent3(c)
  }
  const pageChange4 = (v: number) => {
    const c = v
    setCurrent4(c)
  }
  const itemRender = (item: any) => {
    return <div>{item.number === 3 ? 'hot' : item.text}</div>
  }
  const pageChange5 = (v: number) => {
    const c = v
    setCurrent3(c)
  }
  return (
    <div className="demo">
      <h2>{translated.basic}</h2>
      <Cell>
        <Pagination
          value={currentPage1}
          total={20}
          pageSize={5}
          onChange={pageChange1}
        />
      </Cell>
      <h2>{translated.simple}</h2>
      <Cell>
        <Pagination
          value={currentPage2}
          total={12}
          pageSize={1}
          mode="simple"
          onChange={pageChange2}
        />
      </Cell>
      <h2>{translated.lite}</h2>
      <Cell>
        <Pagination
          value={currentPage2}
          total={12}
          pageSize={1}
          mode="lite"
          onChange={pageChange2}
        />
      </Cell>
      <h2>{translated.ellipse}</h2>
      <Cell>
        <Pagination
          value={currentPage3}
          total={125}
          itemSize={2}
          ellipse
          onChange={pageChange3}
        />
      </Cell>
      <h2>{translated.custom}</h2>
      <Cell>
        <Pagination
          value={currentPage4}
          total={500}
          itemSize={5}
          onChange={pageChange4}
          prev={<Left />}
          next={<Right />}
          itemRender={itemRender}
        />
      </Cell>
      <h2>{translated.uncontrolled}</h2>
      <Cell>
        <Pagination
          defaultValue={15}
          total={500}
          pageSize={10}
          itemSize={3}
          onChange={pageChange5}
        />
      </Cell>
    </div>
  )
}

export default PaginationDemo
