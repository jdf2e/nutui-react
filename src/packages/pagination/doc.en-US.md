# Pagination

## Intro

When the amount of data is too much, use pagination to separate the data.

## Install

```tsx
import { Pagination } from '@nutui/nutui-react';
```

## Demo

### Basic Usage

When the current page number is bound by `value`, the component is in a controlled state, and the paging display depends on the incoming `value`, which is generally used with `onChange`. When it does not need to be controlled, the current page number can be specified through `defaultValue`

:::demo

```tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage1, setCurrentPage1] = useState(1)
  const pageChange1 = (v: number) => {
    const c = v
    setCurrentPage1(c)
  }
  return (
    <Pagination
      value={currentPage1}
      total={25}
      pageSize={5}
      onChange={pageChange1}
    />
  )
}
export default App;
```

:::

### Simple mode

Pagination can be switched to simple mode with simple mode attribute, and pagination cann't display specific page buttons.

:::demo

```tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage2, setCurrentPage2] = useState(1)
  const pageChange2 = (v: number) => {
    const c = v
    setCurrentPage2(c)
  }
  return (
    <Pagination
      value={currentPage2} 
      total={12}
      pageSize={1}
      mode="simple" 
      onChange={pageChange2} 
    />
  )
}
export default App;
```

:::

### Show ellipses

The ellipses button will display after with force-ellipses attribute, click it can jump quickly 

:::demo

```tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage3, setCurrentPage3] = useState(1)
  const pageChange3 = (v: number) => {
    const c = v
    setCurrentPage3(c)
  }
  return (
    <Pagination
      value={currentPage3}
      total={125}
      itemSize={2}
      ellipse
      onChange={pageChange3}
    />
  )
}
export default App;
```

:::

### Custom Button

Pass in a custom method through itemRender, parameters: { number: "page number", text: "page text", active: "active page" } 


:::demo

```tsx
import React, { useState } from 'react'
import { Pagination} from '@nutui/nutui-react'; 
import { Left, Right } from '@nutui/icons-react';

const App = () => {
  const [currentPage4, setCurrentPage4] = useState(1)
  const pageChange4 = (v: number) => {
    const c = v
    setCurrentPage4(c)
  }
  const itemRender = (page: any) => {
    return <div>{page.number === 3 ? 'hot' : page.text}</div>
  }
  return (
    <Pagination
      value={currentPage4}
      total={500}
      itemSize={5}
      onChange={pageChange4}
      itemRender={itemRender} 
      prev={<Left />}
      next={<Right />}
    />
  )
}
export default App;
```

:::

### Uncontrolled Mode

:::demo

```tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react'; 

const App = () => {
  const pageChange5 = (v: number) => {
    console.log(v)
  }
  return (
    <Pagination
      defaultValue={15}
      total={500}
      pageSize={10}
      itemSize={3}
      onChange={pageChange5}
    />
  )
}
export default App;
```

:::

## Pagination

### Props

| Property | Description | Type | Default |
| --- | --- | --- | --- |
| value | current page number, controlled value | `number` | `-` |
| defaultValue | default page number, uncontrolled value | `number` | `1` |
| mode | Display mode  | `multi` \| `simple` | `multi` |
| prev | Customize previous page button content | `ReactNode` | `Previous` |
| next | Customize next page button content | `ReactNode` | `Next` |
| total | total | `number` | `50` |
| pageSize | records per page | `number` | `10` |
| itemSize | number of pages displayed | `number` | `5` |
| ellipse | Whether to show ellipsis | `boolean` | `false` |
| itemRender | Used to customize page number content | `(page: {number, text}) => ReactNode` | `-` |
| onChange | when the page number changes | `(value) => void` | `-` |

## Theming

### CSS Variables

The component provides the following CSS variables, which can be used to customize styles. Please refer to [ConfigProvider component](#/en-US/component/configprovider).

| Name | Description | Default |
| --- | --- | --- |
| \--nutui-pagination-color | font color  |  `$primary-color` |
| \--nutui-pagination-font-size | font size | `$font-size-2` |
| \--nutui-pagination-item-border-color | border color  | `$gray9` |
| \--nutui-pagination-disable-color | disable color | `$disable-color` |
| \--nutui-pagination-disable-background-color | disable background color | `$gray10` |
| \--nutui-pagination-item-border-width | border width | `1px` |
| \--nutui-pagination-item-border-radius | border radius  | `2px` |
| \--nutui-pagination-prev-next-padding | padding  | `0 11px` |