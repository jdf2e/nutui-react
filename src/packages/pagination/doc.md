#  Pagination 分页

### 介绍
    
当数据量较多时，采用分页的形式分隔长列表。
    
### 安装
``` javascript
import { Pagination } from '@nutui/nutui-react';
```    

### 基础用法
通过modelValue来绑定当前页码时，组件为受控状态，分页显示取决于传入的modelValue，一般搭配onChange使用。
不需要受控时，可通过defaultCurrentPage指定当前页码
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage1, setCurrentPage1] = useState(1)
  const pageChange1 = (v: any) => {
    let c = v
    setCurrentPage1(c)
  }
  return (
    <Pagination
      modelValue={currentPage1}
      totalItems="25"
      itemsPerPage="5"
      onChange={pageChange1}
    />
  )
}
export default App;
```
:::
### 简单模式
将 mode 设置为 "simple" 来切换到简单模式，此时分页器不会展示具体的页码按钮。
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage2, setCurrentPage2] = useState(1)
  const pageChange2 = (v: any) => {
    let c = v
    setCurrentPage2(c)
  }
  return (
    <Pagination
      modelValue={currentPage2} 
      pageCount={12} 
      mode="simple" 
      onChange={pageChange2} 
    />
  )
}
export default App;
```
:::

### 显示省略号
设置 force-ellipses 后会展示省略号按钮，点击后可以快速跳转。
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage3, setCurrentPage3] = useState(1)
  const pageChange3 = (v: any) => {
    let c = v
    setCurrentPage3(c)
  }
  return (
    <Pagination
      modelValue={currentPage3}
      totalItems="125"
      showPageSize="3"
      forceEllipses
      onChange={pageChange3}
    />
  )
}
export default App;
```
:::
### 自定义按钮
通过pageNodeRender传入自定义方法，入参数为page:{ number:页数, text:"文本", active:"选中状态" } 
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination,Icon } from '@nutui/nutui-react';

const App = () => {
  const [currentPage4, setCurrentPage4] = useState(1)
  const pageChange4 = (v: any) => {
    let c = v
    setCurrentPage4(c)
  }
  const pageNodeRender = (page: any) => {
    return <div>{page.number == 3 ? 'hot' : page.text}</div>
  }
  return (
    <Pagination
      modelValue={currentPage4}
      totalItems="500"
      showPageSize="5"
      onChange={pageChange4}
      pageNodeRender={pageNodeRender} 
      prevText={<Icon name="left"/>} 
      nextText={<Icon name="right"/>}
    />
  )
}
export default App;
```
:::
    
## API
    
### Props
    
| 参数           | 说明                             | 类型                      | 默认值            |
| -------------- | -------------------------------- | ------------------------- | ----------------- |
| modelValue     | 当前页码                         | Number                    | -                 |
| defaultValue   | 当前页码                         | Number                    | 1                 |
| mode           | 显示模式,可选值为：multi，simple | String                    | multi             |
| prevText       | 自定义上一页按钮内容             | String \| React.ReactNode | 上一页            |
| nextText       | 自定义下一页按钮内容             | String \| React.ReactNode | 下一页            |
| pageCount      | 总页数                           | String \| Number          | 传入/根据页数计算 |
| totalItems     | 总记录数                         | String \| Number          | 0                 |
| itemsPerPage   | 每页记录数                       | String \| Number          | 10                |
| showPageSize   | 显示的页码个数                   | String \| Number          | 5                 |
| forceEllipses  | 是否显示省略号                   | Boolean                   | false             |
| pageNodeRender | 用于自定义页码的结构             | (page) => React.ReactNode | -                 |
    
### Events
    
| 事件名   | 说明           | 回调参数 |
| -------- | -------------- | -------- |
| onChange | 页码改变时触发 | value    |
