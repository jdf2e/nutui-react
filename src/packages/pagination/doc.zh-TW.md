#  Pagination 分頁

### 介紹
    
當數據量較多時，採用分頁的形式分隔長列表。
    
### 安裝
``` javascript
// react
import { Pagination } from '@nutui/nutui-react';
```    

### 基礎用法
通過modelValue來綁定當前頁碼時，組件為受控狀態，分頁顯示取決於傳入的modelValue，一般搭配onChange使用。
不需要受控時，可通過defaultCurrentPage指定當前頁碼
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage1, setCurrentPage1] = useState(1)
  const pageChange1 = (v: any) => {
    const c = v
    setCurrentPage1(c)
  }
  return (
    <Pagination
      modelValue={currentPage1}
      totalItems="25"
      itemsPerPage="5"
      onChange={pageChange1}
    />
  )
}
export default App;
```
:::
### 簡單模式
將 mode 設置為 "simple" 來切換到簡單模式，此時分頁器不會展示具體的頁碼按鈕。
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage2, setCurrentPage2] = useState(1)
  const pageChange2 = (v: any) => {
    const c = v
    setCurrentPage2(c)
  }
  return (
    <Pagination
      modelValue={currentPage2} 
      pageCount={12} 
      mode="simple" 
      onChange={pageChange2} 
    />
  )
}
export default App;
```
:::

### 顯示省略號
設置 force-ellipses 後會展示省略號按鈕，點擊後可以快速跳轉。
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination } from '@nutui/nutui-react';

const App = () => {
  const [currentPage3, setCurrentPage3] = useState(1)
  const pageChange3 = (v: any) => {
    const c = v
    setCurrentPage3(c)
  }
  return (
    <Pagination
      modelValue={currentPage3}
      totalItems="125"
      itemSize="3"
      ellipse
      onChange={pageChange3}
    />
  )
}
export default App;
```
:::
### 自定義按鈕
通過pageNodeRender傳入自定義方法，入參數為page:{ number:頁數, text:"文本", active:"選中狀態" }
:::demo
``` tsx
import React, { useState } from 'react'
import { Pagination} from '@nutui/nutui-react'; 
import { Left, Right } from '@nutui/icons-react-taro';

const App = () => {
  const [currentPage4, setCurrentPage4] = useState(1)
  const pageChange4 = (v: any) => {
    const c = v
    setCurrentPage4(c)
  }
  const pageNodeRender = (page: any) => {
    return <div>{page.number === 3 ? 'hot' : page.text}</div>
  }
  return (
    <Pagination
      modelValue={currentPage4}
      totalItems="500"
      itemSize="5"
      onChange={pageChange4}
      pageNodeRender={pageNodeRender} 
      prev={<Left />}
      next={<Right />}
    />
  )
}
export default App;
```
:::
    
## API
    
### Props
    
| 屬性 | 說明 | 類型 | 預設值           |
| -------------- | -------------------------------- | ------------------------- | ----------------- |
| modelValue     | 當前頁碼                         | number                    | -                 |
| defaultValue   | 當前頁碼                         | number                    | `1`                 |
| mode           | 顯示模式,可選值為：multi，simple  | string                    | `multi`             |
| prev       | 自定義上一頁按鈕內容             | string \| ReactNode | `上一頁`            |
| next       | 自定義下一頁按鈕內容             | string \| ReactNode | `下一頁`            |
| pageCount      | 總頁數                           | string \| number          | 傳入/根據頁數計算 |
| totalItems     | 總記錄數                         | string \| number          | `0`                 |
| itemsPerPage   | 每頁記錄數                       | string \| number          | `10`                |
| itemSize   | 顯示的頁碼個數                   | string \| number          | `5`                 |
| ellipse  | 是否顯示省略號                   | boolean                   | `false`             |
| pageNodeRender | 用於自定義頁碼的結構             | (page) => ReactNode | -                 |
    
### Events
    
| 事件名稱 | 說明 | 回調參數     |
| -------- | -------------- | -------- |
| onChange | 页码改变时触发 | `value`    |


## 主題定制

### 樣式變量

組件提供了下列 CSS 變量，可用於自定義樣式，使用方法請參考 [ConfigProvider 組件](#/zh-CN/component/configprovider)。

| 名稱 | 默認值 |
| --- | --- |
| --nutui-pagination-color | `$primary-color` |
| --nutui-pagination-font-size | `$font-size-2` |
| --nutui-pagination-item-border-color | `#e4e7eb` |
| --nutui-pagination-disable-color | `$disable-color` |
| --nutui-pagination-disable-background-color | `#f7f8fa` |
| --nutui-pagination-item-border-width | `1px` |
| --nutui-pagination-item-border-radius | `2px` |
| --nutui-pagination-prev-next-padding | `0 11px` |
