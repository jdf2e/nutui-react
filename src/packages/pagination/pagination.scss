.nut-pagination {
  display: flex;
  flex-direction: row;
  font-size: $pagination-font-size;
  color: $pagination-color;
}

.nut-pagination-prev,
.nut-pagination-item,
.nut-pagination-next {
  height: 39px;
  min-width: 39px;
  flex-shrink: 0;
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: $pagination-font-size;
  color: $pagination-color;
  background: $white;
  border-radius: $pagination-item-border-radius;
  border: $pagination-item-border-width solid $pagination-item-border-color;
  cursor: pointer;
}

.nut-pagination-prev,
.nut-pagination-item {
  border-right-width: 0;
}

.nut-pagination-prev,
.nut-pagination-next {
  padding: $pagination-prev-next-padding;
}

.nut-pagination-contain {
  display: flex;
  flex-direction: row;
}

.nut-pagination-item-active {
  color: $white;
  border-width: 0;
  background-color: $pagination-active-background-color;
}

.nut-pagination-item-disabled,
.nut-pagination-next-disabled,
.nut-pagination-prev-disabled {
  color: $pagination-disable-color;
  background-color: $pagination-disable-background-color;
  cursor: not-allowed;
}

.nut-pagination-simple {
  height: 39px;
  width: 124px;
  line-height: 39px;
  text-align: center;
  font-size: $pagination-font-size;
  color: $pagination-color;
}

.nut-pagination-simple-border {
  border-right: $pagination-item-border-width solid
    $pagination-item-border-color;
}

.nut-pagination-lite {
  width: $pagination-lite-width;
  height: $pagination-lite-height;
  display: flex;
  flex-direction: row;
  color: $color-primary-text;
  background-color: $pagination-lite-background-color;
  border-radius: $pagination-lite-radius;
}

.nut-pagination-lite-active,
.nut-pagination-lite-default {
  display: flex;
  align-items: center;
  justify-content: center;
  width: $pagination-lite-height;
  height: $pagination-lite-height;
  font-size: $font-size-small;
  color: $pagination-lite-color;
  font-weight: 400;
}

.nut-pagination-lite-active {
  border-radius: $pagination-lite-radius;
  font-size: $font-size-small;
  background-color: $pagination-lite-active-background-color;
}
