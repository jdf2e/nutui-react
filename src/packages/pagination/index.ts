import { Pagination } from './pagination'

export type { PaginationProps } from './pagination'
export default Pagination
