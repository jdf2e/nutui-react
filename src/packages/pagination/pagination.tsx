import React, { FunctionComponent, useEffect, useState } from 'react'
import classNames from 'classnames'
import { useConfig } from '@/packages/configprovider'

export interface PaginationProps {
  defaultValue: number
  current: number
  mode: 'multi' | 'simple'
  prev: React.ReactNode
  next: React.ReactNode
  total: string | number
  pageSize: string | number
  itemSize: string | number
  ellipse: boolean
  itemRender: (page: any) => React.ReactNode
  onChange: (currPage: number) => void
  style?: React.CSSProperties
  className?: string
}

const defaultProps = {
  defaultValue: 1,
  mode: 'multi',
  prev: null,
  next: null,
  total: '0',
  pageSize: '10',
  itemSize: '5',
  ellipse: false,
  className: '',
  itemRender: (item: any) => {
    return item.text
  },
} as PaginationProps
export const Pagination: FunctionComponent<
  Partial<PaginationProps> &
    Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'>
> = (props) => {
  const { locale } = useConfig()
  const { children } = { ...defaultProps, ...props }
  const {
    current,
    mode,
    prev,
    next,
    total,
    pageSize,
    itemSize,
    onChange,
    ellipse,
    itemRender,
    defaultValue,
    className,
    ...rest
  } = props

  const [currentPage, setCurrent] = useState(1)
  const [pages, setPages] = useState<unknown[]>([])
  const [countRef, setCountRef] = useState(0)
  const classPrefix = 'nut-pagination'
  // 计算页面的数量
  const computedCountRef = () => {
    const num = Math.ceil(Number(total) / Number(pageSize))
    return Number.isNaN(num) ? 1 : Math.max(1, num)
  }

  // 生成pages数组，用来遍历
  const computedPages = (_currentPage?: number, _countRef?: number) => {
    if (mode === 'simple') return []
    const items = []
    const pageCount = _countRef || countRef // 总的页面数量
    const pageSize = Number(itemSize) // 展示的页面个数
    const _current = _currentPage || Number(currentPage) // 当前页
    let startPage = 1
    let endPage = pageCount
    const partialShow = pageCount > pageSize
    if (partialShow) {
      // 选中的page在展示的page中间
      startPage = Math.max(_current - Math.floor(pageSize / 2), 1)
      endPage = startPage + pageSize - 1
      if (endPage > pageCount) {
        endPage = pageCount
        startPage = endPage - pageSize + 1
      }
    }
    // 遍历生成数组
    for (let i = startPage; i <= endPage; i++) {
      const page = setPage(i, i, _current === i)
      items.push(page)
    }
    // 判断是否有折叠
    if (partialShow && pageSize > 0 && ellipse) {
      if (startPage > 1) {
        const prevPage = setPage(startPage - 1, '...')
        items.unshift(prevPage)
      }
      if (endPage < pageCount) {
        const nextPage = setPage(endPage + 1, '...')
        items.push(nextPage)
      }
    }
    return items
  }
  // 点击选择page
  const selectPage = (curPage: number, isSelect: boolean) => {
    if (curPage > countRef || curPage < 1) return
    // 是否传入current
    if (!('current' in props)) {
      setCurrent(Number(curPage))
      if (curPage !== currentPage) {
        setPages(computedPages(curPage))
      }
    }
    if (isSelect) onChange && onChange(curPage)
  }
  // set page 对象
  const setPage = (number: any, text: any, active?: any) => {
    return { number, text, active }
  }

  useEffect(() => {
    // 初始化 判断是否传入值 计算 当前页，总页数，生成子节点
    let currentValue = props.defaultValue || 1
    if ('current' in props) {
      currentValue = Number(props.current)
    }
    const pageCount = computedCountRef()
    setCurrent(currentValue)
    setCountRef(pageCount)
    setPages(computedPages(currentValue, pageCount))
  }, [])

  if ('current' in props) {
    const current = props.current ? Number(props.current) : 1
    if (current !== currentPage) {
      setCurrent(current)
      setPages(computedPages(Number(current)))
    }
  }
  return (
    <div className={classNames(classPrefix, className)} {...rest}>
      <div
        className={classNames(
          `${classPrefix}__prev`,
          mode === 'multi' ? '' : 'simple-border',
          currentPage === 1 ? 'disabled' : ''
        )}
        onClick={(e) => selectPage(Number(currentPage) - 1, true)}
      >
        {prev || locale.pagination.prev}
      </div>
      {mode === 'multi' ? (
        <div className={`${classPrefix}__contain`}>
          {pages.map((item: any, index: number) => {
            return (
              <div
                key={`${index}pagination`}
                className={classNames(
                  `${classPrefix}__item`,
                  item.active ? 'active' : ''
                )}
                onClick={(e) =>
                  !item.active ? selectPage(item.number, true) : ''
                }
              >
                {itemRender ? itemRender(item) : item.text}
              </div>
            )
          })}
        </div>
      ) : (
        ''
      )}
      {mode === 'simple' ? (
        <div className={`${classPrefix}__contain`}>
          <div className={`${classPrefix}__simple`}>
            {currentPage}/{countRef}
          </div>
        </div>
      ) : (
        ''
      )}
      <div
        className={classNames(
          `${classPrefix}__next`,
          Number(currentPage) >= countRef ? 'disabled' : ''
        )}
        onClick={(e) => selectPage(Number(currentPage) + 1, true)}
      >
        {next || locale.pagination.next}
      </div>
    </div>
  )
}

Pagination.defaultProps = defaultProps
Pagination.displayName = 'NutPagination'
