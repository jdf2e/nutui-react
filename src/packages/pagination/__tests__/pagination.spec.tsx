import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Pagination } from '../pagination'

test('should match snapshot', () => {
  const { asFragment } = render(<Pagination total={25} pageSize={5} />)
  expect(asFragment()).toMatchSnapshot()
})

test('should render items', async () => {
  const { container } = render(<Pagination total={25} pageSize={5} />)
  expect(container.querySelectorAll('.nut-pagination__item')).toHaveLength(5)
})
test('should render simple mode', async () => {
  const { container } = render(
    <Pagination total={12} pageSize={1} mode="simple" />
  )
  expect(container.querySelectorAll('.nut-pagination__item')).toHaveLength(0)
  expect(container.querySelectorAll('.nut-pagination__simple')).toHaveLength(1)
})
test('should render ellipse and should emit change event after clicking ellipse option', async () => {
  const { container, getByText } = render(
    <Pagination total={125} itemSize={3} ellipse />
  )
  expect(container.querySelectorAll('.nut-pagination__item')).toHaveLength(4)
  fireEvent.click(getByText('...'))
  expect(container.querySelectorAll('.nut-pagination__item')).toHaveLength(5)
  expect(
    container.querySelectorAll('.nut-pagination__item')[1]
  ).toHaveTextContent('3')
})

test('should emit change event after clicking visible option', async () => {
  const { container, getByText } = render(
    <Pagination total={25} pageSize={5} defaultValue={1} />
  )
  const next = getByText('下一页')
  fireEvent.click(next)

  expect(container.querySelectorAll('.nut-pagination__item')).toHaveLength(5)
  expect(container.querySelectorAll('.nut-pagination__item')[1]).toHaveClass(
    'active'
  )
})

test('should not emit change event after clicking disable option', async () => {
  let flag = false
  const pageChange = (v: number) => {
    flag = true
  }
  const { container, getByText } = render(
    <Pagination
      total={25}
      pageSize={5}
      defaultValue={1}
      onChange={pageChange}
    />
  )
  const prev = getByText('上一页')
  fireEvent.click(prev)

  expect(container.querySelectorAll('.nut-pagination__item')).toHaveLength(5)
  expect(flag).toBeFalsy()
})

test('should render custom content correctly', () => {
  const itemRender = (page: any) => {
    return <div>{page.number === 3 ? 'hot' : page.text}</div>
  }
  const { container, getByText } = render(
    <Pagination
      total={25}
      pageSize={5}
      defaultValue={1}
      itemRender={itemRender}
      prev="pre"
      next="next"
    />
  )
  expect(getByText('pre')).toHaveTextContent('pre')
  expect(getByText('next')).toHaveTextContent('next')
  expect(
    container.querySelectorAll('.nut-pagination__item')[2]
  ).toHaveTextContent('hot')
})

test('test controlled mode', () => {
  let current = 2
  const pageChange = (v: number) => {
    current = v
  }
  const { container, getByText } = render(
    <Pagination
      current={current}
      total={25}
      pageSize={5}
      onChange={pageChange}
    />
  )
  expect(container.querySelector('.active')).toHaveTextContent('2')
  const page = getByText('4')
  fireEvent.click(page)
  expect(current).toEqual(4)
})

test('test uncontrolled mode', () => {
  let current = 0
  const pageChange = (v: number) => {
    current = v
  }
  const { container, getByText } = render(
    <Pagination
      defaultValue={2}
      total={25}
      pageSize={5}
      onChange={pageChange}
    />
  )
  expect(container.querySelector('.active')).toHaveTextContent('2')
  const page = getByText('4')
  fireEvent.click(page)
  expect(current).toEqual(4)
  expect(container.querySelector('.active')).toHaveTextContent('4')
})
