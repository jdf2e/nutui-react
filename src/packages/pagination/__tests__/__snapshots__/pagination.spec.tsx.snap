// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="nut-pagination"
  >
    <div
      class="nut-pagination-prev disabled"
    >
      上一页
    </div>
    <div
      class="nut-pagination-contain"
    >
      <div
        class="nut-pagination-item active"
      >
        1
      </div>
      <div
        class="nut-pagination-item"
      >
        2
      </div>
      <div
        class="nut-pagination-item"
      >
        3
      </div>
      <div
        class="nut-pagination-item"
      >
        4
      </div>
      <div
        class="nut-pagination-item"
      >
        5
      </div>
    </div>
    <div
      class="nut-pagination-next"
    >
      下一页
    </div>
  </div>
</DocumentFragment>
`;
